### PR TITLE
Added fast_sync mode.

### PR DESCRIPTION
**Pull request by kgpayne**
## Problem

'Fast Sync' is a feature of pipelinewise, not singer. This means folks using this target outside of pipelinewise do not benefit (*cough*Meltano). This PR enables fast-sync, configurable by setting "fast_sync": true in your config.json file.

WARNING: fast sync must be supported and enabled by both tap and target for it to work.

## Proposed changes

Piping records between tap and target using STDOUT is slow for tables with large numbers of rows. Therefore I have implemented record batching, with a collection of records serialized to disk and a pointer written into a new BatchRecord. As I have no direct control over the Singer spec, this batch record appears as a standard RECORD, using config and a reserved key (__fast_sync_message__) to identify these new messages.

My intention is that `FULL_TABLE` and first-run `LOG_BASED` replication happens in batch _if_ `fast_sync` is enabled. The behaviour of `INCREMENTAL` tables remains unaffected, in part because there is no obvious branch-point to wedge a `if batch:` statement and we don't need it for testing. Having reviewed the code since this PR, it should be possible to implement `fast_sync` for incremental tables by removing the `if replication_method == 'INCREMENTAL'` check on line 176 of `common.py` and updating the checkpointing logic on lines 260 to 277. Will add this in a future PR.

## Types of changes

What types of changes does your code introduce to PipelineWise?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [x] Description above provides context of the change
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Unit tests for changes (not needed for documentation changes)
- [ ] CI checks pass with my changes
- [ ] Bumping version in `setup.py` is an individual PR and not mixed with feature or bugfix PRs
- [ ] Commit message/PR title starts with `[AP-NNNN]` (if applicable. AP-NNNN = JIRA ID)
- [ ] Branch name starts with `AP-NNN` (if applicable. AP-NNN = JIRA ID)
- [ ] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [ ] Relevant documentation is updated including usage instructions

### CODE REVIEW

File: tap_mysql/sync_strategies/common.py [hunk 1]
```diff
@@ -144,59 +161,122 @@ def sync_query(cursor, catalog_entry, state, select_sql, columns, stream_version
     LOGGER.info('Running %s', query_string)
     cursor.execute(select_sql, params)
 
-    row = cursor.fetchone()
-    rows_saved = 0
-
     database_name = get_database_name(catalog_entry)
 
+    md_map = metadata.to_map(catalog_entry.metadata)
+    stream_metadata = md_map.get((), {})
```
Comment (NiallRees) [inline tap_mysql/sync_strategies/common.py:162]: What does this do? 🤔

**Comment by NiallRees** [comment]
Again for future record, please could you outline what this change implements and what it doesn't implement? Does it know to use fast sync for the first run of an incremental or log based sync for example, and then switch to incremental/log based for subsequent syncs?

**Review by NiallRees — APPROVED** [review]

File: tap_mysql/sync_strategies/common.py (hunk 1 above)
Comment (kgpayne) [reply tap_mysql/sync_strategies/common.py:162]: It was pulled from the old lines 166 and 167 (below). Think it grabs the metadata object that came in an earlier message.

Comment (NiallRees) [reply tap_mysql/sync_strategies/common.py:162]: Ah ok cool.

**Comment by kgpayne** [comment]
> Again for future record, please could you outline what this change implements and what it doesn't implement? Does it know to use fast sync for the first run of an incremental or log based sync for example, and then switch to incremental/log based for subsequent syncs?

Thanks 😅  I hadn't ticked the "Relevant documentation is updated including usage instructions" for a reason 🤣

Jokes aside, I haven't actually implemented fast-sync for incremental tables in this PR. There is no explicit "if first time" logic for incremental, so I deferred the implementation till later to crack on with an end-to-end prototype. Having looped back, it looks like incremental just pulls the last synced key from state. If there is no key, it selects back to the beginning of time. So it should actually be very easy to add to fast-sync (as per the updates I made to the description 👆). Will work on a second PR in the morning.

**Review by alanmcruickshank — APPROVED** [review]
Based on my (albeit limited) understanding of what's going on, this looks pretty sensible.

**Pull request merged**